Code generation documentation: prose style

tokencolour opened the pull request:
Extends examples on autowrap, ufuncify, binary_function,
ufuncify, codegen, make_routine and the code printers.
Fixes #8889

tokencolour committed:
Code generation documentation: prose style

Extends examples on autowrap, ufuncify, binary_function,
ufuncify, codegen, make_routine and the code printers.
Fixes #8889

moorepants commented on doc/src/tutorial/codegen_explainatory.rst:
```diff
@@ -0,0 +1,453 @@
+>>> binary_func = autowrap(expr)
+>>> binary_func(1, 4, 2)
+-1.0
+
```
I think correct restructuredtext for code blocks is:

```
The callable returned from autowrap() is a binary python function, not a 
SymPy object. For example::

   >>> from sympy.abc import x, y, z
   >>> from sympy.utilities.autowrap import autowrap
   >>> expr = ((x - y + z)**(13)).expand()
   >>> binary_func = autowrap(expr)
   >>> binary_func(1, 4, 2)
   -1.0
```

moorepants commented:
This is looking really good. I think that the tutorial is the right place for this.

tokencolour commented:
Thank you, I'll do the change you just mentioned. I suppose all the code blocks should be formatted this way.

moorepants commented:
Yes they should. You should build the documentation on your computer to make sure things render correctly.

tokencolour committed:
Code generation documentation: prose style

Extends examples on autowrap, ufuncify, binary_function
codegen, make_routine and the code printers.
Fixes #8889
Changed to the correct reStructuredText for code blocks
as suggested by @moorepants

tokencolour commented:
@moorepants I have corrected the format for code blocks. Kindly review it.

moorepants commented on doc/src/tutorial/codegen_explainatory.rst:
```diff
@@ -0,0 +1,458 @@
+    >>> m=symbols('m', integer=True)
+    >>> i = Idx('i', m)
+    >>> a,omega= symbols('a, omega')
+    >>> qho=autowrap(Eq(y[i], psi_n(0, x[i], m, omega)), tempdir='/home/thd/Documents/tmp')
```
The doctests will run this on Travis. You should either use a directory in the current working directory or one in `/tmp/`. This may not work as is and it is better to make it more general.

tokencolour replied:
Changed this!

moorepants commented on doc/src/tutorial/codegen_explainatory.rst:
```diff
@@ -0,0 +1,458 @@
+expression. If the main expression needs to call a specialized function it should 
+be put in the ``helpers`` iterable. Autowrap will then make sure that the
+compiled main expression can link to the helper routine. Items should
+be tuples with (<funtion_name>, <sympy_expression>, <arguments>). It is mandatory 
```
funtion>function

tokencolour replied:
Changed!

moorepants commented on doc/src/tutorial/codegen_explainatory.rst:
```diff
@@ -0,0 +1,458 @@
+    >>> from sympy import symbols
+    >>> from sympy.physics.hydrogen import R_nl
+    >>> a,r = symbols('a,r')
+    >>> psi_nl=R_nl(1,0,a,r)
```
Please go through your code snippets and make them PEP8 compatible, for example this line should be:

`>>> psi_n1 = R_n1(1, 0, a, r)`

tokencolour replied:
Changed all instances of this mistake. If you could spot more errors, it would be great, since I don't know much about PEP8 compatibilty.

moorepants commented on doc/src/tutorial/codegen_explainatory.rst:
```diff
@@ -0,0 +1,458 @@
+
+Another method available at the ``autowrap`` level is ``binary_function``. It returns 
+a sympy function. The advantage is that we can have very fast functions as compared
+to SymPy speeds. An illustration::
```
This could use a bit better explanation. Why would this create something fast?

moorepants replied:
Oh, I see. These are compiled functions but have sympy attributes and methods. In particular, `evalf` will evaluate faster than a standard SymPy object.

tokencolour replied:
`evalf` will evaulate faster for a `binary_function` than for a standard Sympy object, wouldn't it?

moorepants commented on doc/src/tutorial/codegen_explainatory.rst:
```diff
@@ -0,0 +1,458 @@
+
+
+While NumPy operations are very efficient for vectorized data but they sometimes incur 
+unnecessary costs when chained together.  
```
What do you mean by this statement?

tokencolour replied:
I had read it [here](http://docs.sympy.org/dev/modules/numeric-computation.html#ufuncify). I did not include the full explanation. Should I?
Chaining i.e. using 2 or more numpy operations together is unoptimized in Numpy as the source states.

moorepants commented on doc/src/tutorial/codegen_explainatory.rst:
```diff
@@ -0,0 +1,458 @@
+It can then depend on projects like Cython or f2py to compile and reconnect that 
+code back up to Python. Fortunately this process is well automated and a SymPy user 
+wishing to make use of this code generation should call the ufuncify function
+``ufuncify`` is the third method available with Autowrap module. 
```
The main point of ufuncify as compared to autowrap is that it allows arrays as arguments and follows NumPy's broadcasting rules.

tokencolour replied:
I've added an explanation. Kindly review it once I push the changes.

moorepants commented on doc/src/tutorial/codegen_explainatory.rst:
```diff
@@ -0,0 +1,458 @@
+code back up to Python. Fortunately this process is well automated and a SymPy user 
+wishing to make use of this code generation should call the ufuncify function
+``ufuncify`` is the third method available with Autowrap module. 
+It basically implies 'Universal functions' and follows an ideology set by Numpy.
```
Maybe link back to the NumPy ufuncs documentation here.

tokencolour replied:
Added!

moorepants commented on doc/src/tutorial/codegen_explainatory.rst:
```diff
@@ -0,0 +1,458 @@
+It basically implies 'Universal functions' and follows an ideology set by Numpy.
+Let us see an example::
+
+    >>> from sympy import *
```
Please avoid using `from sympy import *`, just import what you need.

moorepants commented on doc/src/tutorial/codegen_explainatory.rst:
```diff
@@ -0,0 +1,458 @@
+Let us see an example::
+
+    >>> from sympy import *
+    >>> init_printing(use_latex='mathjax')
```
This should have args to enable pretty printing not mathjax. In fact, just call it with no arguments.

moorepants commented on doc/src/tutorial/codegen_explainatory.rst:
```diff
@@ -0,0 +1,458 @@
+    >>> from sympy.utilities.autowrap import ufuncify
+    >>> from sympy.utilities.lambdify import lambdify
+    >>> fn_numpy=lambdify(x, expr,'numpy')
+    >>> fn_fortran=ufuncify([x], expr)
```
Are you sure the default is fortran here? I thought the default was numpy.

jksuom replied:
Is the spelling of the file name correct?

tokencolour replied:
Oh, yes! The default for ufuncify is numpy. I need to change this to add the backend=f2py arg.
Though, I can't guess why numpy is beating numpy.

tokencolour replied:
@jksuom I need to change the file name. Thanks for pointing out.

moorepants commented on doc/src/tutorial/codegen_explainatory.rst:
```diff
@@ -0,0 +1,458 @@
+Various flags to ``codegen`` let you modify things like the project name
+and global variables.
+
+    name_expr : tuple, or list of tuples
```
The following seems to be a copy and paste from the codgen doc string. Is that necessary since you can find it elsewhere? I prefer your style above where you introduced different args and than showed an example.

tokencolour replied:
I have changed this.

moorepants commented on doc/src/tutorial/codegen_explainatory.rst:
```diff
@@ -0,0 +1,458 @@
+    >>> expr=R_nl(3,y,x,6)
+    >>> r=make_routine('my_routine',expr)
+    >>> [arg.result_var for arg in r.results]
+    [result₅₁₄₂₃₄₁₆₈₁₃₉₇₇₁₉₄₂₈]
```
What's up with this printing?

tokencolour replied:
I can't help it. I tried init_printing without any arguments as you suggested. I still get a similar result. Please help me with this.

moorepants commented on doc/src/tutorial/codegen_explainatory.rst:
```diff
@@ -0,0 +1,458 @@
+Javascript (``jscode``), Mathematica (``mathematica_code``), Octave/Matlab (``octave_code``), 
+Python (print_python, which is actually more like a lightweight version 
+of codegen for Python, and ``lambdarepr``, which supports many libraries 
+(like NumPy), and theano, if that counts (``theano_code``).The code printers 
```
Theano counts.

moorepants commented:
I really like this. Nice job!

One thing, I feel like the whole document is in the wrong order. I'd personally like to see the summary explanation with the layers at the top and the go through the layers bottom up in stead of top down. What do you think? Is is easier to understand what is going on by thinking about autowrap first or by thinking about the code printers first?

moorepants commented:
- It would be nice to show some examples of matlab/octave code generation.
- It would be nice to show one reasonably complicated expression and show the printing output for each language we support.

tokencolour commented:
Thank you. The document is in line with the wiki-note on code generation by @asmeurer, you told me about. Changing the order of the document seems like a nice idea. I'll try to do so.
And if we add examples with a complicated expression in all code printers, will there be a need to add more examples on matlab/octave code generation? (since the API of all code printers is the same). And thank you for the review.

tokencolour commented:
Also `from sympy.utilities.autowrap import autowrap` doesn't work for me on Sympy live, neither does importing ufuncify and binary_function.
![screenshot from 2015-12-19 13 17 38](https://cloud.githubusercontent.com/assets/9028362/11912073/0cd2e866-a653-11e5-82cc-e30cf419454b.png)
 What should I do to make it work?

tokencolour committed:
[WIP]Code generation documentation: prose style

Added the changes suggested by @moorepants
Also changed the order of the doc
Not yet included the example of a complicated expression
Fixes #8889

asmeurer commented:
Autowrap won't work on SymPy live. It requires the ability to write code to a file and compile it, and the app engine doesn't have any compilers installed (and apparently you can't even use the subprocess module anyway).

tokencolour commented:
So that means, I should add doctest: +SKIP in the examples of `autowrap`, because those are some of those places where the checks are failing.
Thank you, @asmeurer

moorepants commented on doc/src/tutorial/codegen_explanatory.rst:
```diff
@@ -0,0 +1,443 @@
+that help use other numeric systems as backend for construction and 
+manipulation of mathematical expressions to speed up the process.
+
+Unconventionally, we will start with a summary, and inspect the building 
```
It is now an intro and not a summary, which is not unconventional.

moorepants commented on doc/src/tutorial/codegen_explanatory.rst:
```diff
@@ -0,0 +1,443 @@
+Unconventionally, we will start with a summary, and inspect the building 
+blocks of the code generation and then the code generation itself.
+
+Summary
```
Summary > Introduction

moorepants commented on doc/src/tutorial/index.rst:
```diff
@@ -17,3 +17,4 @@
    solvers.rst
    matrices.rst
    manipulation.rst
+   codegen_explainatory.rst
```
codegen.rst is sufficient. What you have now is misspelled.

tokencolour committed:
Code generation documentation: prose style

Changed file name.
Added examples for octave and mathematica code
printers. Also added the example of a complicated
expression.
Fixes #8889

tokencolour commented:
@moorepants , thank you for the review. I have added more examples and renamed the file. Kindly take a look at it.

tokencolour committed:
Code generation documentation: prose style

Added comments to skip doctests.
Fixes #8889

tokencolour commented:
And I can't figure out why the checks have failed this time. Please help me.

jksuom commented:
mpmath.org had closed the connection. I've restarted the job.

tokencolour commented:
Thanks, @jksuom .

moorepants commented on doc/src/tutorial/codegen.rst:
```diff
@@ -0,0 +1,516 @@
+    >>> expr = R_nl(3,y,x,6)
+    >>> r = make_routine('my_routine',expr)
+    >>> [arg.result_var for arg in r.results]   # doctest: +SKIP
+    [result₅₁₄₂₃₄₁₆₈₁₃₉₇₇₁₉₄₂₈]
```
This printing is correct.

moorepants commented on doc/src/tutorial/codegen.rst:
```diff
@@ -0,0 +1,516 @@
+    >>> from sympy import symbols, init_printing
+    >>> init_printing()
+    >>> x, y = symbols('x y')
+    >>> expr = R_nl(3,y,x,6)
```
PEP8 > Spaces after the commas.

Please have a look at PEP8 and/or run a linter over your code to check PEP8 compliance. Please go through this whole PR and fix the PEP8 errors.

moorepants commented on doc/src/tutorial/codegen.rst:
```diff
@@ -0,0 +1,516 @@
+
+    end subroutine
+
+Using the argument 'args' along with it changes argument sequence::
```
If you use double ticks ``, the literals will print correctly. So args should have two ticks on each side.

bjodah commented:
This is really nice addition! I could not find any remarks other than the PEP8 ones already put forward. Great job :+1: !

moorepants commented:
Yes, I'll merge after the PEP8 things are fixed. This looks great.

moorepants commented on doc/src/tutorial/codegen.rst:
```diff
@@ -0,0 +1,516 @@
+==========================
+Code Generation from Sympy
```
One more comment, this should just be "Code Generation".

tokencolour committed:
Code generation documentation: prose style

Tried to make code PEP8 compatible.
Changed Title.
Fixes #8889

tokencolour commented:
@bjodah Thank you!

tokencolour commented:
@moorepants I read about PEP8 compatibility and I have tried to fix those errors. Kindly have a look at it. Sorry for being ignorant and thank you for the review.

moorepants commented on doc/src/tutorial/codegen.rst:
```diff
@@ -0,0 +1,516 @@
+    >>> t = IndexedBase('t', shape=(len_y,))
+    >>> Dy = IndexedBase('Dy', shape=(len_y-1,))
+    >>> i = Idx('i', len_y-1)
+    >>> e = Eq(Dy[i], (y[i+1]-y[i])/(t[i+1]-t[i]))
```
It is preferable to use whitespace around operators:

`>>> e = Eq(Dy[i], (y[i+1] - y[i]) / (t[i+1] - t[i])`

https://www.python.org/dev/peps/pep-0008/#other-recommendations

tokencolour replied:
Shouldn't it be this?
`>>> e = Eq(Dy[i], (y[i+1]-y[i]) / (t[i+1]-t[i])`

moorepants replied:
I think that spaces are recommended around operators except in slices/indexing, `[i+1]`. You can remove the spaces for operators in a string that have precedence. See the link I sent.

moorepants replied:
This stuff is sorta trivial, but just make it more readable with some spaces. Right now if has no spaces and the readability is not so great.

tokencolour replied:
I've added the spaces the way you suggested, it looks best that way. Kindly see the updated PR.

tokencolour committed:
Code generation documentation: prose style

More improvements on PEP8 compatibility
Fixes #8889

tokencolour commented:
@moorepants Please take a look at this.

tokencolour committed:
Code generation documentation: prose style

Fixes #8889
improves on PEP8 compatibility

moorepants commented:
@tokencolour Thanks for this contribution. This will be very helpful for our users.

Merging.

tokencolour commented:
Thank you!

asmeurer commented on doc/src/tutorial/codegen.rst:
```diff
@@ -0,0 +1,516 @@
+like abs(x) -> fabs(x) (for C).
+
+The code printers don't print optimal code in many cases. 
+An example of this is powers in C. x**2 prints as pow(x, 2) instead of x*x. 
```
These should have been wrapped as code. I think they will render as italics.

tokencolour replied:
Changed!

asmeurer commented on doc/src/tutorial/codegen.rst:
```diff
@@ -0,0 +1,516 @@
+------------------------------
+This is where the meat of code generation is, the translation of SymPy
+expressions to specific code.Supported languages are C (``ccode``), Fortran 95 (``fcode``), 
+Javascript (``jscode``), Mathematica (``mathematica_code``), Octave/Matlab (``octave_code``), 
```
You should use :func:\`jscode\` for these, so that they link to their docs (you may need to do :func:\`sympy.printing.jscode.jscode\`).

tokencolour replied:
`jscode`, `octave_code`, `theano_code` aren't rendering properly, because they haven't been included in `printing.rst` doc in modules. What should be done about them?

moorepants replied:
You can include the autodoc for them in the printing.rst file.

asmeurer commented:
Sorry I didn't get to look at this before it was merged, but I do have some comments. @tokencolour will you be able to fix them? If not, I will fix them.

asmeurer commented on doc/src/tutorial/codegen.rst:
```diff
@@ -0,0 +1,516 @@
+    >>> x, y, z = symbols('x y z')
+    >>> mat = Matrix([x, y, z]).T
+    >>> known_mat = MatrixSymbol('K', 1, 3)
+    >>> Assignment(known_mat, mat)  # doctest: +SKIP
```
Why did you skip these doctests? You should only need to skip the functions that compile code (autowrap).

tokencolour replied:
This is because this particular line shows an error on SymPy live
```
>>> Assignment(known_mat, mat)
Traceback (most recent call last):
  File "/base/data/home/apps/s~sympy-live-hrd/45.380685391411058699/sympy/sympy/interactive/printing.py", line 28, in _displayhook
    print(stringify_func(arg))
  File "/base/data/home/apps/s~sympy-live-hrd/45.380685391411058699/sympy/sympy/printing/pretty/pretty.py", line 1917, in pretty
    return pp.doprint(expr)
  File "/base/data/home/apps/s~sympy-live-hrd/45.380685391411058699/sympy/sympy/printing/pretty/pretty.py", line 50, in doprint
    return self._print(expr).render(**self._settings)
  File "/base/data/home/apps/s~sympy-live-hrd/45.380685391411058699/sympy/sympy/printing/printer.py", line 257, in _print
    return getattr(self, printmethod)(expr, *args, **kwargs)
  File "/base/data/home/apps/s~sympy-live-hrd/45.380685391411058699/sympy/sympy/printing/pretty/pretty.py", line 136, in _print_Relational
    op = prettyForm(' ' + xsym(e.rel_op) + ' ')
  File "/base/data/home/apps/s~sympy-live-hrd/45.380685391411058699/sympy/sympy/printing/pretty/pretty_symbology.py", line 441, in xsym
    op = _xsym[sym]
KeyError: ':='
```
Though, the following are giving correct results.
```
>>> Assignment(known_mat, mat).lhs
K
>>> Assignment(known_mat, mat).rhs
[x  y  z]
```
I am sorry I forgot to bring your attention to this before. What should I do about this?

moorepants replied:
This code is working fine for me:

```
Python 3.5.1 |Continuum Analytics, Inc.| (default, Dec  7 2015, 11:16:01) 
Type "copyright", "credits" or "license" for more information.

IPython 4.0.1 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: from sympy import symbols, MatrixSymbol, Matrix

In [2]: from sympy.printing.codeprinter import Assignment

In [3]: x, y, z = symbols('x y z')

In [4]: mat = Matrix([x, y, z]).T

In [5]: known_mat = MatrixSymbol('K', 1, 3)

In [6]: Assignment(known_mat, mat)
Out[6]: K := Matrix([[x, y, z]])
```

It shouldn't be skipped.

What are you doing with SymPy live? You should test these locally with you Python/SymPy installation.

tokencolour replied:
I tested them locally, it works fine on my local machine. But it failed tests on Travis. Anyway, I'll not skip this doctest and see what happens and maybe, then bring it to your attention again.

moorepants replied:
Pleaes remove all of the skip statements and send a new pull request so we can see what is failing on Travis. In general, the solution for failures on Travis is never going to be to skip the test. That defeats the purpose of testing.

tokencolour replied:
I'll keep that in mind.

asmeurer replied:
SymPy Live runs the release version and this has changed since the release. In general, you should not rely on SymPy Live for SymPy development.

tokencolour replied:
I'll keep that in mind.

asmeurer commented on doc/src/tutorial/codegen.rst:
```diff
@@ -0,0 +1,516 @@
+
+
+Autowrap
+--------
```
Do you need an empty line here?

tokencolour replied:
Changed!

asmeurer commented on doc/src/tutorial/codegen.rst:
```diff
@@ -0,0 +1,516 @@
+
+
+While NumPy operations are very efficient for vectorized data but they sometimes incur 
+unnecessary costs when `chained together <http://docs.sympy.org/dev/modules/numeric-computation.html#ufuncify>`_  .
```
You should use a cross reference here, not linking directly to the docs.

tokencolour replied:
Changed. Added a label in numeric-computation.rst to cross reference.

asmeurer commented on doc/src/tutorial/codegen.rst:
```diff
@@ -0,0 +1,516 @@
+
+
+The lambdify function translates SymPy expressions into Python functions, 
+leveraging a variety of numerical libraries.By default lambdify relies 
```
Missing space after period.

tokencolour replied:
Changed.

asmeurer commented on doc/src/tutorial/index.rst:
```diff
@@ -17,3 +17,4 @@
    solvers.rst
    matrices.rst
    manipulation.rst
+   codegen.rst
```
I am -0.5 to including this in the tutorial. The tutorial is designed for beginners to SymPy, but this document is intended for users who are already familiar with the basics of SymPy. I would rather see this elsewhere (maybe modules, or special_topics).

moorepants replied:
Do we need an advanced tutorial?

This doesn't really go with any particular module because the modules in it are spread around. It could be listed as a general topic instead of a tutorial. It is pretty much like a tutorial but it borders on just being an explanation of the modules too. But this can also be considered the beginner's tutorial for the code generation aspects of SymPy. I feel like the code gen stuff is more related to the core of sympy functionality than other extra modules, like mechanics, etc.

tokencolour commented:
I would be happy to make these amendments. I deleted the branch, though. It would be a new pull request.

moorepants commented:
@tokencolour Where are you making the changes? In a new PR?

tokencolour commented:
@moorepants No, I found that I did not delete the branch. I misremembered this and I am sorry I did not tell it.

tokencolour commented:
@asmeurer  and @moorepants, Thank you for your comments. Kindly take a look at this new PR. Where should I keep the doc?
